org/postgresql/jdbc2/CallableQueryKey.java

```diff
@@ -0,0 +1,45 @@
+/*-------------------------------------------------------------------------
+*
+* Copyright (c) 2015, PostgreSQL Global Development Group
+*
+*
+*-------------------------------------------------------------------------
+*/
+package org.postgresql.jdbc2;
+
+/**
+ * Serves as a cache key for callable statements.
+ * For regular statements, just sql string is used as a key.
+ */
+class CallableQueryKey {
+    public final String sql;
+
+    public CallableQueryKey(String sql)
+    {
+        this.sql = sql;
+    }
+
+    @Override
+    public String toString()
+    {
+        return sql;
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o) return true;
+        if (!(o instanceof CallableQueryKey)) return false;
+
+        CallableQueryKey that = (CallableQueryKey) o;
+
+        return !(sql != null ? !sql.equals(that.sql) : that.sql != null);
```

Comment:
On 28 June 2015 at 11:29, Sehrope Sarkuni <notifications@github.com> wrote:

> In org/postgresql/jdbc2/CallableQueryKey.java
> <https://github.com/pgjdbc/pgjdbc/pull/319#discussion_r33423791>:
>
> > +
> > +    @Override
> > +    public String toString()
> > +    {
> > +        return sql;
> > +    }
> > +
> > +    @Override
> > +    public boolean equals(Object o)
> > +    {
> > +        if (this == o) return true;
> > +        if (!(o instanceof CallableQueryKey)) return false;
> > +
> > +        CallableQueryKey that = (CallableQueryKey) o;
> > +
> > +        return !(sql != null ? !sql.equals(that.sql) : that.sql != null);
>
> The double negative is a bit hard to follow. How about:
>
> return sql == null ? that.sql == null : sql.equals(that.sql);
>
> , or:
>
> if (sql==null) return that.sql == null;
> return sql.equals(that.sql);
>
>
Either one is more readable...

Dave